### PR TITLE
Refactor FilterDependencies to use LINQ groups

### PR DESCRIPTION
**Pull request by Shivansps**
This is more clear to read, it also finishes adding "<=" support if it is set externally(it is for some mods on the nebula db), ill not add support for it in the UI because to make it work properly for every case it would need a complete rewrite the dependency system.

Knet 2.0 will need a complete re-write of the dependency system.

### CODE REVIEW

File: Knossos.NET/Models/Mod.cs [hunk 1]
```diff
@@ -348,127 +348,221 @@ public List<string> GetModFlagList(bool overrideSettings = false)
...
-                                    {
-                                        temp.Remove(eqOrHigh);
-                                    }
-                                });
-                                revisions.ForEach(revs =>
+                                    higher.Remove(h);
+                                }
+                            });
+                        });
+
```
Comment (JohnAFernandez) [inline Knossos.NET/Models/Mod.cs:459]: I didn't see a section where `==` dependencies filter out `~` dependencies, as you said in the main comment above.

File: Knossos.NET/Models/Mod.cs [hunk 2]
```diff
@@ -348,127 +348,221 @@ public List<string> GetModFlagList(bool overrideSettings = false)
...
+                                    rev.Remove(r);
+                                }
+                            });
+                        });
+                        rev.ForEach(re => {
+                            //Remove any "lower" that can be sastified by this rev
+                            lower.ToList().ForEach(l =>
                             {
-                                equalOrHigher.ForEach(eqOrHigh =>
+                                if (SemanticVersion.SastifiesDependency(l.version, re.version!.Replace("~", "")))
```
Comment (JohnAFernandez) [inline Knossos.NET/Models/Mod.cs:475]: Does this properly filter out `l` if `l` is `<=1.0.1` and `re` is `~1.0.2` ?

Same for higher dependencies, would `h` be filtered out when `h` is `>=1.0.2` and `re` is `1.0.1`?

File: Knossos.NET/Models/Mod.cs [hunk 3]
```diff
@@ -348,127 +348,221 @@ public List<string> GetModFlagList(bool overrideSettings = false)
...
+                        lower.ForEach(lo => {
+                            //Remove any "higher" that can be sastified by this lower
+                            higher.ToList().ForEach(h =>
+                            {
+                                if (SemanticVersion.SastifiesDependency(h.version, lo.version!.Replace("<=", "")))
+                                {
+                                    higher.Remove(h);
+                                }
+                            });
+                        });
```
Comment (JohnAFernandez) [inline Knossos.NET/Models/Mod.cs:510]: Hmmmm ... the way I'm reading this, if `l` is `<=1.1.0` and `h` is `>=1.0.2`, `h` is filtered out.  But that allows us to use `1.0.1`.  Is that desired?

**Review by JohnAFernandez — COMMENTED** [review]
Definitely have some questions on this.

File: Knossos.NET/Models/Mod.cs (hunk 2 above)
Comment (Shivansps) [reply Knossos.NET/Models/Mod.cs:475]: On the first one no, that would be a conflict, "<=1.0.1" cant be sastified with "1.0.2" and "~1.0.2" cant be sastified by "1.0.1". So they both will be in

On the second one mmm "~1.0.1" should be sastified by "1.0.2" what would leave >=1.0.2 out thats definately wrong

File: Knossos.NET/Models/Mod.cs (hunk 3 above)
Comment (Shivansps) [reply Knossos.NET/Models/Mod.cs:510]: Yes thats the flaw with the current dependency system

File: Knossos.NET/Models/Mod.cs (hunk 1 above)
Comment (Shivansps) [reply Knossos.NET/Models/Mod.cs:459]: Equals are checked on https://github.com/KnossosNET/Knossos.NET/blob/8ff6e6b167b177d1a61e84f1e2bc0f9dfbfffc8d/Knossos.NET/Models/Mod.cs#L433

File: Knossos.NET/Models/Mod.cs (hunk 2 above)
Comment (Shivansps) [reply Knossos.NET/Models/Mod.cs:475]: mmm no, i just ran a test (i have another branch were i did the changes to the UI to test)

image

And it seem to correctly mark that case as a conflict

**Comment by Shivansps** [comment]
I see no point in merging this now that the dependency system will be changed sooner rather than later.

**Pull request closed without merging**